### PR TITLE
rclcpp::Duration in RobotTrajectory

### PR DESCRIPTION
**Pull request by tylerjw**
Signed-off-by: Tyler Weaver <maybe@tylerjw.dev>

### Description

Use the strong type `rclcpp::Duration` in `RobotTrajectory` and opportunistically convert member functions to free functions.

### CODE REVIEW

**Comment by tylerjw** [comment]
Ideally I'd like to make the robot trajectory a composed type inside a single deque instead of two parallel deques and then move as much as possible out into free functions. I am only getting started with this PR cleaning this type up.

File: moveit_core/trajectory_processing/src/ruckig_traj_smoothing.cpp [hunk 1]
```diff
@@ -220,7 +220,7 @@ bool RuckigSmoothing::runRuckig(robot_trajectory::RobotTrajectory& trajectory,
   trajectory.unwind();
 
   // Initialize the smoother
-  double timestep = trajectory.getAverageSegmentDuration();
+  double timestep = robot_trajectory::average_segment_duration(trajectory).seconds();
```
Comment (henningkayser) [inline moveit_core/trajectory_processing/src/ruckig_traj_smoothing.cpp:223]: Why not use the provided member function instead? We shouldn't mix these two different paradigms. If you are planning to make RobotTrajectory a pure data type with free functions, we can talk about that. Here this change is unnecessarily confusing.

**Review by henningkayser — COMMENTED** [review]
I really like the move away from using double for time durations. We should probably do this for all of the codebase at some point. I can relate to your idea of making RobotTrajectory a composed type, but then this should happen in one refactoring, or at least we should have an API ready for discussion before merging these changes. IMO, it would make more sense to implement the composed type and then deprecate the old class altogether.

File: moveit_core/robot_trajectory/include/moveit/robot_trajectory/robot_trajectory.h [hunk 2]
```diff
@@ -144,32 +145,57 @@ class RobotTrajectory
     return waypoints_.front();
   }
 
-  const std::deque<double>& getWayPointDurations() const
+  std::deque<rclcpp::Duration> const& getDurationFromPreviousDeque() const
```
[the remark this answers is not in the document]
Comment (tylerjw) [reply moveit_core/robot_trajectory/include/moveit/robot_trajectory/robot_trajectory.h:148]: The problem is that you can't overload return types to deprecate a function where you want to change the return type.

File: moveit_core/robot_trajectory/include/moveit/robot_trajectory/robot_trajectory.h [hunk 3]
```diff
@@ -144,32 +145,57 @@ class RobotTrajectory
...
+                   [](auto const& e) { return e.seconds(); });
+    return ret;
+  }
+
   /** @brief  Returns the duration after start that a waypoint will be reached.
    *  @param  The waypoint index.
    *  @return The duration from start; returns overall duration if index is out of range.
    */
-  double getWayPointDurationFromStart(std::size_t index) const;
+  rclcpp::Duration getWayPointDurationFromStartAt(std::size_t index) const;
```
[the remark this answers is not in the document]
Comment (tylerjw) [reply moveit_core/robot_trajectory/include/moveit/robot_trajectory/robot_trajectory.h:166]: Long term I'd like to make all these free functions. 

Maybe I should just design that and we can discuss that plan instead of this incremental PR.

The reason I made this PR was to separate the idea of going from double -> rclcpp::Duration from a larger data-structure change.

File: moveit_core/robot_trajectory/include/moveit/robot_trajectory/robot_trajectory.h [hunk 4]
```diff
@@ -380,11 +464,17 @@ class RobotTrajectory
...
-  std::deque<double> duration_from_previous_;
+  std::deque<rclcpp::Duration> duration_from_previous_;
   rclcpp::Clock clock_ros_;
 };
 
 /** @brief Operator overload for printing trajectory to a stream */
 std::ostream& operator<<(std::ostream& out, const RobotTrajectory& trajectory);
 
+/// \brief Calculate the total duration of a trajectory
+rclcpp::Duration total_duration(RobotTrajectory const& trajectory);
```
[the remark this answers is not in the document]
Comment (tylerjw) [reply moveit_core/robot_trajectory/include/moveit/robot_trajectory/robot_trajectory.h:475]: I'd like to change the norm to be free functions and not Classes with state. Free functions are easier to reason about being correct, easier to test, easier to compose, and easier to make generic and DRY.

Changing the whole code-base to free functions that operate on structs at once is an impossible task, however I could reasonably do it one object at a time.

File: moveit_core/trajectory_processing/src/ruckig_traj_smoothing.cpp (hunk 1 above)
Comment (tylerjw) [reply moveit_core/trajectory_processing/src/ruckig_traj_smoothing.cpp:223]: The reason I picked these two functions and changed them to functions is that you can't deprecate a return type of a function with an overload of the same name, similar to the places where I added an `At` suffix or `Deque` suffix.

File: moveit_core/robot_trajectory/include/moveit/robot_trajectory/robot_trajectory.h [hunk 5]
```diff
@@ -116,7 +117,7 @@ class RobotTrajectory
 
   const moveit::core::RobotState& getWayPoint(std::size_t index) const
   {
-    return *waypoints_[index];
+    return *waypoints_.at(index);
```
Comment (ChrisThrasher) [inline moveit_core/robot_trajectory/include/moveit/robot_trajectory/robot_trajectory.h:120]: This change seems unrelated to using `rclcpp::Duration` and deserves its own conversation in a separate issue or pull request. Once we widen the API in this way, we cannot narrow it again without breaking the API so we must be very careful about such changes.

File: moveit_core/robot_trajectory/include/moveit/robot_trajectory/robot_trajectory.h (hunk 2 above)
Comment (ChrisThrasher) [inline moveit_core/robot_trajectory/include/moveit/robot_trajectory/robot_trajectory.h:148]:
```suggestion
  const std::deque<rclcpp::Duration>& getDurationFromPreviousDeque() const
```
There's no need to sprinkle in stylistic changes. Lets keep style consistent until we can enforce const style with clang-format-14. Same goes for a few other places in this PR. See https://github.com/ros-planning/moveit2/issues/1431

File: moveit_core/robot_trajectory/include/moveit/robot_trajectory/robot_trajectory.h [hunk 6]
```diff
@@ -183,50 +209,86 @@ class RobotTrajectory
...
+    return addSuffixWayPoint(state, rclcpp::Duration::from_seconds(dt));
+  }
+
   /**
    * \brief Add a point to the trajectory
    * \param state - current robot state
    * \param dt - duration from previous
    */
-  RobotTrajectory& addSuffixWayPoint(const moveit::core::RobotStatePtr& state, double dt)
+  RobotTrajectory& addSuffixWayPoint(const moveit::core::RobotStatePtr& state, rclcpp::Duration dt)
```
Comment (ChrisThrasher) [inline moveit_core/robot_trajectory/include/moveit/robot_trajectory/robot_trajectory.h:228]: Is `rclcpp::Duration` the same size as `double`? We ought to be sure about that before we continue to pass it around by value.

File: moveit_core/robot_trajectory/include/moveit/robot_trajectory/robot_trajectory.h [hunk 7]
```diff
@@ -237,9 +299,16 @@ class RobotTrajectory
...
-  RobotTrajectory& append(const RobotTrajectory& source, double dt, size_t start_index = 0,
+  RobotTrajectory& append(const RobotTrajectory& source, rclcpp::Duration dt, size_t start_index = 0,
                           size_t end_index = std::numeric_limits<std::size_t>::max());
 
+  [[deprecated("Use of double for duration is deprecated, please pass in rclcpp::Duration")]] RobotTrajectory&
+  append(const RobotTrajectory& source, double dt, size_t start_index = 0,
+         size_t end_index = std::numeric_limits<std::size_t>::max())
+  {
+    return append(source, rclcpp::Duration::from_seconds(dt), start_index, end_index);
+  }
```
Comment (ChrisThrasher) [inline moveit_core/robot_trajectory/include/moveit/robot_trajectory/robot_trajectory.h:310]: Why is there a mix between some functions being defined in this header and some functions merely being declared in it?

**Review by ChrisThrasher — CHANGES_REQUESTED** [review]

File: moveit_core/robot_trajectory/include/moveit/robot_trajectory/robot_trajectory.h (hunk 6 above)
Comment (tylerjw) [reply moveit_core/robot_trajectory/include/moveit/robot_trajectory/robot_trajectory.h:228]: Internally it is an int64_t representing nanoseconds. I think that's the same size as `double`.

File: moveit_core/robot_trajectory/include/moveit/robot_trajectory/robot_trajectory.h (hunk 7 above)
Comment (tylerjw) [reply moveit_core/robot_trajectory/include/moveit/robot_trajectory/robot_trajectory.h:310]: That is a problem that already exists here and deserves its own cleanup.

File: moveit_core/robot_trajectory/include/moveit/robot_trajectory/robot_trajectory.h (hunk 2 above)
Comment (tylerjw) [reply moveit_core/robot_trajectory/include/moveit/robot_trajectory/robot_trajectory.h:148]: I'll go make sure I clean these out. I have a pattern of doing this without thinking about it because I've been writing all my new code this way recently.

**Comment by tylerjw** [comment]
Closing as this is abandoned unfinished work. I created an issue here to track this if we decide to come back and do this: https://github.com/ros-planning/moveit2/issues/1728
